port: Use memalign or _aligned_malloc only, when available, fallback to malloc.

winspool opened the pull request:
This fixes also the usage of "_WIN32" to decide, 
if the MSVC extensions _aligned_malloc / _aligned_free are available

That bug breaks other Compiler on Windows. (OpenWatcom as Example).

-- 
Regards ... Detlef

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Enhanced build configuration to automatically detect and enable advanced aligned memory allocation methods.
  - Introduced support for optional sanitizers to improve error detection during builds.

- **Refactor**
  - Refined the logic for memory allocation and deallocation, ensuring more robust cross-platform behavior.
  - Improved header inclusion logic for aligned memory allocation methods.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

nmoinvaz commented on CMakeLists.txt:
```diff
@@ -431,6 +434,21 @@ if(HAVE_ALIGNED_ALLOC)
+endif()
+set(CMAKE_REQUIRED_DEFINITIONS)
+
+check_symbol_exists(_aligned_malloc malloc.h HAVE__ALIGNED_MALLOC)
```
Two underscores. Is that intentional?

winspool replied:
Yes, the function starts with an underscore.

Without the second underscore, the difference between the
C11 aligned allocation macro and the 
MS aligned allocation macro is only one character
and can easily mixed:
(HAVE_ALIGNED_ALLOC and HAVE_ALLIGNED_MALLOC) compared to
(HAVE_ALIGNED_ALLOC and HAVE__ALLIGNED_MALLOC)


But of course, when the project wants macros with only one underscore, please let me know.

winspool commented:
When a similar check for cpuid is added, OpenWatcom can build "zlib-ng":
[0000_check_for_cpuid_support.txt](https://github.com/user-attachments/files/18644761/0000_check_for_cpuid_support.txt)


Some TODO items for OpenWatcom:
* Make the OW internal atomic support available
* Make aligned allocation support available
* Make cpuid support available

ccawley2011 commented on zutil_p.h:
```diff
@@ -23,20 +26,26 @@ static inline void *zng_alloc(size_t size) {
     void *ptr;
     return posix_memalign(&ptr, 64, size) ? NULL : ptr;
-#elif defined(_WIN32)
+#elif defined(HAVE__ALIGNED_MALLOC)
```
It might be worth moving this to the top just in case there's a platform that supports both `_aligned_malloc` and `aligned_alloc`/`posix_memalign`, to make sure that the correct `free` function is used.

winspool replied:
Oh yeah.
I changed the detection strategy
for this Revision to reduce indention
(Posix and c11.
Check gnu ext. only, when c11 ist absent
Check MS ext only, when gnu ext. ist absent)

That's better to read, but i forgot to handle the free function.

I will prepare an updated Patch.

ccawley2011 commented on zutil_p.h:
```diff
@@ -23,20 +26,26 @@ static inline void *zng_alloc(size_t size) {
     return (void *)memalign(64, size);
+#else
+    /* Fallback to use a normal allocation and hope the best */
+    /* This works only, when the default C compression code is used */
```
From what I can tell, zlib-ng does ensure that allocated memory is aligned at a higher level, so it might be worth clarifying that in the comments.

winspool replied:
I know only about the alignments required for SSE/AVX.

ccawley2011 replied:
What I meant is that zlib-ng currently includes padding when allocating, which means it can make sure that the buffer is aligned even if the allocation callback doesn't guarantee it (since it might be provided by users of the library rather than using the default implementation).

https://github.com/zlib-ng/zlib-ng/blob/develop/deflate.c#L246
https://github.com/zlib-ng/zlib-ng/blob/develop/inflate.c#L181

mtl1979 commented:
Needs rebase, fail with `Static Analysis / Clang` should already be fixed.

Dead2 commented on zutil_p.h:
```diff
@@ -10,8 +10,11 @@
 #elif defined(__FreeBSD__)
```
**nitpick (non-blocking):** Since you removed the __APPLE__ below in block choosing malloc type in favor of using a fallback, perhaps also remove the above check for __APPLE__ as well.

You could also remove the whole first check there, and just add a comment in the fallback that `HAVE_POSIX_MEMALIGN` and `HAVE_ALIGNED_ALLOC` as well as standard malloc are all handled by the fallback. Seems redundant that both the first and the last block are the same here.

winspool replied:
Ok.

I will prepare a follow-up patch to clean it up
after this change is committed.

mtl1979 commented on zutil_p.h:
```diff
@@ -10,8 +10,11 @@
+#elif defined(HAVE_MEMALIGN) || defined(HAVE__ALIGNED_MALLOC)
 #  include <malloc.h>
+#else
+/* Fallback, when no other aligned allocation was found */
```
Nitpick: Maybe "aligned allocation function" would be better...

winspool committed:
ports: Use memalign or _aligned_malloc, when available. Fallback to malloc

Using "_WIN32" to decide,
if the MSVC extensions _aligned_malloc / _aligned_free are available
is a bug that breaks other Compiler on Windows. (OpenWatcom as Example)

Regards ... Detlef

winspool commented:
I updated the patch to current git
and included the comments from @Dead2 and @mtl1979.

Please let me know, what else i can do to get this patch committed.

Dead2 reviewed (approved):
LGTM

pps83 commented:
IMO, better not to bring this messy complexity and do it ffmpeg/x264 way: they completely abandoned these and simply align memory themselves. Not sure who of them does, or maybe both. If the code needs aligned memory and there is no routines you need to fallback to manual alignment: could as well keep using it for all to make it simple. cost of manual alignment is zero.

```c++
malloc64:
    uint8_t *buf = (uint8_t*)malloc(size + 63 + sizeof(void**));
    if (buf)
    {
        uint8_t *aligned = buf + 63 + sizeof(void**);
        aligned -= (intptr_t)aligned & 63;
        *((void**)(aligned- sizeof(void **))) = buf;
        return aligned;
    }
    

free64:
    if(p)
        free( *(((void**)p) - 1));
```
something like that

https://github.com/zlib-ng/zlib-ng/commit/902abe731495d20c740a1623820fe8e219f697ea

Dead2 commented:
@pps83 Unless I missed something, and I don't think I did, all allocs should already get checked for alignment and if not already aligned it will manually align them. See #1713.

The reason we always check for alignment is that the zlib api allows applications giving it pointers to its preferred alloc/free functions, and we cannot know whether those will supply us with aligned memory.
So, in fact you are right, we could do away with using the aligned alloc functions entirely, if we want to. It would only waste up to 64bytes of ram at the start of the buffer, instead of like today 64 bytes at the end of the buffer. The effective difference to us should be zero.
